Fixed requirements.yml role installation to avoid concurrent job's fa…

Roviluca opened the pull request:
##### SUMMARY
Fixed requirements.yml role installation to avoid concurrent job's failures
    
When performing a project update roles were not downloaded and were downloaded at job execution time, this caused problems accessing and decrypting files updated by other concurrent jobs.

The behavior now is:
- update roles during scm_update
-  perform galaxy-install --force only when scm_clean option is true

##### ISSUE TYPE
 - Bugfix Pull Request

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 2.0.1
```

Commit:
Fixed requirements.yml role installation to avoid concurrent job's failures

When performing a project update roles were not downloaded and were downloaded at job execution time, this caused problems accessing and decrypting files updated by other concurrent jobs.

The behavior now is:
- update roles during scm_update
- perform galaxy-install --force only when scm_clean option is true

wenottingham reviewed (changes requested):
This will not work in any clustered environment where the only checkout happens at job run time.

What SCM are you using?

Roviluca commented:
I went trough trying to resolve ( and i did in my env ) and never described my problem properly, I'll remediate now.

I'm using git as an scm and pointing to branches in requirements.yml file. The infrastructure that made me go check the code-base is built on a single node with containers version 1.0.5.32.
The playbook invoked in my job run via job template is mostly made by attributes and invokes all the roles that contains plain and vaulted files.

The runs are started partially at the same time but more often are started with a little delay with forks value set to 10. Project's "Update Revision on Launch" has been disabled.
I get with a percentage of errors that goes from 10% to 20% with this kind of messages:
> "msg": "Could not find or access 'files/xxx.crt'\nSearched in:\n\t/var/lib/awx/projects/_10__base/roles/ansible_role_xyz/files/xxx.crt\n\t/var/lib/awx/projects/_10__base/roles/ansible_role_xyz/tasks/files/xxx.crt\n\t/var/lib/awx/projects/_10__base/plays/files/xxx.crt"}

>"changed": false, "msg": "Destination /var/lib/awx/projects/_10__base/roles/ansible_role_xyz/files does not exist"

By looking at the code i noticed that requirements are forced at job run and that [ansible-galaxy --force states that will recreate the roles](https://docs.ansible.com/ansible/devel/reference_appendices/galaxy.html#force).

Talking about the solution I think i rebased while you were checking ( my apologies ) and changed the force condition to `{{ scm_clean|ternary('--force','') }}` is it fine for you this way?

wenottingham commented:
(Copying from IRC, just in case)

Checkouts are not guaranteed to be present when the job runs, as they're only run on one node when you do a project update. That is why a checkout is done immediately before running a job - to ensure the content is there.

Restricting ansible-galaxy calls to the triggered project update will mean the roles won't be there in certain cases, such as the job running on a node that didn't run the project update.

wenottingham commented:
The current logic is this, for the checkout that happens at job run time:
- check current git revision
- break if it is already checked out
- if not, check it out and store it in `scm_result`
Then:
- check for role requirements
- if present, run `ansible-galaxy install ...` if `scm_result` is not defined (i.e., when *no* update to the scm was made)
- if present, run `ansible-galaxy install --force` if `scm_result` is defined (i.e., when an update to the SCM was made as part of this checkout)

This should limit the calls with `--force` to project checkouts where the version on disk at the start of the checkout was not the version that was required to run the job - in this case, --force is likely needed, since the requirements have changed.

Roviluca committed:
Merge branch 'devel' into devel

Roviluca commented:
I've found the origin of my errors.
Since i wanted to  ensure that my cloned repositories are always clean and updated I have enabled clean and delete on update at project level and this causes my problems.

Do you think it's worth to create some sort of isolation like create a different folder each time a project is updated with delete on launch?

wenottingham commented:
I'm a bit confused - if clean/delete on update is set, then the non-force `ansible-galaxy install` command should always trigger. Is it not?

Or is the actual clean/delete_on_update process causing the concurrent job problem?

Roviluca commented:
I think the responsible is the delete on update since it deletes everything at the beginning of project update.

Let's put down the flow:
(1) first job started
(1) delete folder
(1) clone repo, install roles
  (2) second job started
  (2) delete folder
**(1) What if job 1 tries to copy files here?**
  (2) clone repo, install roles
(1) first role finish
  (2) second role finish

wenottingham commented:
closing in the context of #3177, would need changed after that
